src/foundry/foundry.js/hooks.d.ts

```diff
@@ -1,830 +1,804 @@
-     * @see {@link SightLayer#restrictVisibility}
+     * A hook event that fires for every Document type before execution of an update workflow. Substitute the Document
+     * name in the hook event to target a specific Document type, for example "preUpdateActor". This hook only fires
+     * for the client who is initiating the update request.
+     *
+     * The hook provides the differential data which will be used to update the Document. Hooked functions may modify
+     * that data or prevent the workflow entirely by explicitly returning false.
+     *
+     * @param document - The Document instance being updated
+     * @param change   - Differential data that will be used to update the document
+     * @param options  - Additional options which modify the update request
+     * @param userId   - The ID of the requesting user, always game.user.id
+     * @returns Explicitly return false to prevent update of this Document
+     * @remarks The name for this hook is dynamically created by joining 'preUpdate' with the type name of the Document.
+     * @remarks This is called {@link Hooks.call}.
+     * @see {@link ClientDatabaseBackend#_preUpdateDocumentArray}
+     * @see {@link TokenDocument#_preUpdateTokenActor}
      */
-    sightRefresh: (layer: SightLayer) => unknown;
+    type PreUpdateDocument<D extends DocumentConstructor = DocumentConstructor> = (
+      document: D,
+      change: DeepPartial<ConstructorParameters<D>>,
+      options: DocumentModificationOptions,
+      userId: string
+    ) => boolean | void;
 
     /**
-     * This is called after the targeted state for a {@link Token} changed.
-     * @param user     - the User that caused the targeted state change
-     * @param token    - the Token for which the targeted state changed
-     * @param targeted - whether the Token is targeted
-     * @remarks This is called by {@link Hooks.callAll}.
-     * @see {@link UserTargets#_hook}
+     * A hook event that fires whenever this Application is rendered.
+     * The hook provides the pending application HTML which will be added to the DOM.
+     * Hooked functions may modify that HTML or attach interactive listeners to it.
+     *
+     * @param app   - The Application instance being rendered
+     * @param html  - The inner HTML of the document that will be displayed and may be modified
+     * @param data  - The object of data used when rendering the application
+     * @typeParam D - the type of the Application data
+     * @typeParam A - the type of the Application
+     * @remarks The name for this hook is dynamically created by joining 'render' with the type name of the Application.
+     * @remarks This is called by {@link Hooks.call}.
+     * @see {@link Application#_render}
      */
-    targetToken: (user: User, token: Token, targeted: boolean) => unknown;
+    type RenderApplication<D = object, A extends Application = Application> = (
+      app: A,
+      html: JQuery,
+      data: D
```

Comment:
It should be possible derive this type from `A`, it should just be `ReturnType<A['getData']>`